### PR TITLE
Jax interface for all devices.

### PR DESCRIPTION
**Pull request by Thenerdstation**
Ported over from #1065 

**Context:**
Previously, only the `default.qubit.jax` device supported the JAX interface. This PR adds JAX interface support to all devices including built-in gradient support for non-default simulators. 

**Description of the Change:**
The recent addition of `jax.experimental.host_callback` allows us to finally do `jax -> numpy -> numpy -> jax` within a `jax.jit` function! This PR adds the needed scaffolding to do it. 

**Benefits:**
Everyone can start using JAX more and I won't take any more excuses.

**Possible Drawbacks:**
None.

**Related GitHub Issues:**
#943 

**TODOs**

* Vmap support

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,22 @@
 
 <h3>New features since last release</h3>
 
+- The JAX interface now supports all devices.
```
Comment (josh146) [inline .github/CHANGELOG.md:5]:
```suggestion
- The JAX interface now supports all devices.
  [(#1076)](https://github.com/PennyLaneAI/pennylane/pull/1076)
```

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,22 @@
...
+  dev = qml.device('cirq.simulator', wires=1)
+  @qml.qnode(dev, interface="jax")
+  def circuit(x):
+      qml.RX(x[1], wires=0)
+      qml.Rot(x[0], x[1], x[2], wires=0)
+      return qml.expval(qml.PauliZ(0))
+  weights = jnp.array([0.2, 0.5, 0.1])
+  print(circuit(weights)) # DeviceArray(...)
+  ```
+  [(#1076)](https://github.com/PennyLaneAI/pennylane/pull/1076)
```
Comment (josh146) [inline .github/CHANGELOG.md:19]:
```suggestion
```

**Review by josh146 — APPROVED** [review]

File: pennylane/tape/interfaces/jax.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    dtype = jnp.float64
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "jax"
+
+    def _execute(self, params, device):
+        # TODO (chase): Add support for more than 1 measured observable.
+        if len(self.observables) != 1:
+            raise ValueError("Only one return type is supported currently")
```
Comment (mariaschuld) [inline pennylane/tape/interfaces/jax.py:85]: Wonder if one should make this more descriptive? Like 
```suggestion
            raise ValueError("The JAX interface currently only supports quantum nodes with a single return type.")
```

File: pennylane/tape/interfaces/jax.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        return_type = self.observables[0].return_type
+        if return_type is not Variance and return_type is not Expectation:
+            raise ValueError(
+                f"Only Variance and Expectation returns are support, given {return_type}"
```
Comment (mariaschuld) [inline pennylane/tape/interfaces/jax.py:89]: Again, maybe mention that this is the JAX interface speaking here, and not general PennyLane?

File: pennylane/tape/interfaces/jax.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    # pylint: disable=attribute-defined-outside-init
+    dtype = jnp.float64
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "jax"
+
+    def _execute(self, params, device):
+        # TODO (chase): Add support for more than 1 measured observable.
```
Comment (mariaschuld) [inline pennylane/tape/interfaces/jax.py:83]: There will be a follow-up PR then?

File: tests/devices/test_default_qubit_jax.py [hunk 6]
```diff
@@ -89,6 +89,26 @@ def circuit(x):
         # Just test that it works and spits our the right value.
         assert jnp.isclose(circuit(p), expected, atol=tol, rtol=0)
 
+    def test_qubit_circuit_different_interface(self, tol):
+        """Test that the device provides the correct
```
Comment (mariaschuld) [inline tests/devices/test_default_qubit_jax.py:93]:
```suggestion
        """Test that the interface provides the correct
```
Was this what you meant?

File: tests/devices/test_default_qubit_jax.py [hunk 7]
```diff
@@ -89,6 +89,26 @@ def circuit(x):
         # Just test that it works and spits our the right value.
         assert jnp.isclose(circuit(p), expected, atol=tol, rtol=0)
 
+    def test_qubit_circuit_different_interface(self, tol):
+        """Test that the device provides the correct
+        result for a simple circuit with a device using a different interface."""
+        if not qml.tape_mode_active():
+            pytest.skip("Tape mode only test")
```
Comment (mariaschuld) [inline tests/devices/test_default_qubit_jax.py:96]: Do we still need this?

File: tests/devices/test_default_qubit_jax.py [hunk 8]
```diff
@@ -89,6 +89,26 @@ def circuit(x):
         # Just test that it works and spits our the right value.
         assert jnp.isclose(circuit(p), expected, atol=tol, rtol=0)
 
+    def test_qubit_circuit_different_interface(self, tol):
+        """Test that the device provides the correct
+        result for a simple circuit with a device using a different interface."""
```
Comment (mariaschuld) [inline tests/devices/test_default_qubit_jax.py:94]: I am a bit confused. Do you mean "Test that a jax QNode returns the correct result when used with a non-jax device"? But would this test be in the right place then?

File: tests/devices/test_default_qubit_jax.py [hunk 9]
```diff
@@ -89,6 +89,26 @@ def circuit(x):
...
+        p = jnp.array(0.543)
+        dev = qml.device("default.qubit.autograd", wires=1)
+
+        @qml.qnode(dev, interface="jax")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliY(0))
+
+        expected = -jnp.sin(p)
+        # Do not test isinstance here since the @jax.jit changes the function
```
Comment (mariaschuld) [inline tests/devices/test_default_qubit_jax.py:107]: May be blind, but does it jit here? I though for this you need an explicit decorator?

File: tests/devices/test_default_qubit_jax.py [hunk 10]
```diff
@@ -436,26 +456,6 @@ def cost(weights):
...
-            qml.CNOT(wires=[0, 1])
-            return qml.expval(qml.PauliZ(0))
-
-        qnode = qml.QNode(circuit, dev, interface="jax", diff_method="parameter-shift")
-        weights = jnp.array([0.1, 0.2])
-
-        with pytest.raises(qml.QuantumFunctionError, match="The JAX interface can only be used with"):
-            qnode(weights)
-
-
```
Comment (mariaschuld) [inline tests/devices/test_default_qubit_jax.py:458]: Nice!

File: tests/tape/interfaces/test_qnode_jax.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+
+	@qml.qnode(dev, interface="jax")
+	def circuit(weights):
+		qml.RX(weights[0], wires=0)
+		qml.RZ(weights[1], wires=1)
+		return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+	weights = jnp.array([0.1, 0.2])
+	val = circuit(weights)
+	assert "DeviceArray" in val.__repr__()
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:33]: Hmm...what would this test? What is a `DeviceArray`?

File: tests/tape/interfaces/test_qnode_jax.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+# limitations under the License.
+"""Unit tests for the JAX interface"""
+import pytest
+jax = pytest.importorskip("jax")
+jnp = pytest.importorskip("jax.numpy")
+import numpy as np
+import pennylane as qml
+from pennylane.tape import JacobianTape, qnode, QNode, QubitParamShiftTape
+
+def test_qnode_intergration():
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:22]: Don't forget to add docstrings to every test function - they are so super useful to understand what the code sets out to actually check!

File: tests/tape/interfaces/test_qnode_jax.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+	@qml.qnode(dev, interface="jax", diff_method="parameter-shift")
+	def circuit(weights):
+		qml.RX(weights[0], wires=0)
+		qml.RZ(weights[1], wires=1)
+		return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+	weights = jnp.array([0.1, 0.2])
+	val = jax.jacrev(circuit)(weights)
+	assert "DeviceArray" in val.__repr__()
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:61]: Also here, what you are testing is not the value, but the return type, right? Maybe mention in test names?

Comment (mariaschuld) [reply tests/tape/interfaces/test_qnode_jax.py:61]: Would one have to test shape/values of the return value as well? I'm just thinking of a situation where the return value is some trivial "zero" or so because something went wrong, but still has the right type...?

File: tests/tape/interfaces/test_qnode_jax.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+	def circuit(weights):
+		qml.RX(weights[0], wires=0)
+		qml.RZ(weights[1], wires=1)
+		return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+	weights = jnp.array([0.1, 0.2])
+	val = jax.grad(circuit)(weights)
+	assert "DeviceArray" in val.__repr__()
+
+@pytest.mark.parametrize("diff_method", ['parameter-shift', 'finite-diff'])
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:75]: Any merit in testing backprop?

File: tests/tape/interfaces/test_qnode_jax.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+        [
+            np.cos(p[1]) * (np.cos(a) * np.cos(p[0]) - np.sin(a) * np.sin(p[0]) * np.sin(p[2])),
+            np.cos(p[1]) * np.cos(p[2]) * np.sin(a)
+            - np.sin(p[1])
+            * (np.cos(a) * np.sin(p[0]) + np.cos(p[0]) * np.sin(a) * np.sin(p[2])),
+            np.sin(a)
+            * (np.cos(p[0]) * np.cos(p[1]) * np.cos(p[2]) - np.sin(p[1]) * np.sin(p[2])),
+        ]
+    )
+    assert np.allclose(res, expected, atol=tol, rtol=0)
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:119]: Maybe this is copied from other interfaces, but if the test checks "unnesting" would one not rather check a tape queue here?

File: tests/tape/interfaces/test_qnode_jax.py [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    new_qnode.construct = construct.__get__(new_qnode, QNode)
+    return new_qnode
+
+
+@pytest.mark.parametrize(
+    "dev_name,diff_method",
+    [("default.mixed", "finite-diff"), ("default.qubit.autograd", "parameter-shift")],
+)
+def test_transform(dev_name, diff_method, monkeypatch, tol):
+    """Test an example transform"""
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:161]: Nice idea!

File: tests/tape/interfaces/test_qnode_jax.py [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+        op2 = qml.RX(weights[1], wires=0)
+        return qml.expval(qml.PauliZ(wires=0))
+
+    weights = np.array([0.32, 0.543])
+    a = np.array(0.5)
+
+    def loss(weights, a):
+        # the following global variable is defined simply for testing
+        # purposes, so that we can easily extract the transformed QNode
+        # for verification.
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:180]: Wait, which global variable?

Comment (mariaschuld) [reply tests/tape/interfaces/test_qnode_jax.py:180]: Oh, `new_circuit`, right?

File: tests/tape/interfaces/test_qnode_jax.py [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+
+    res = loss(weights, a)
+
+    grad = jax.grad(loss, argnums=[0, 1])(weights, a)
+    assert len(grad) == 2
+    assert grad[0].shape == weights.shape
+    assert grad[1].shape == a.shape
+
+    # compare against the expected values
+    tol = 1e-5
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:202]: Maybe not so important, but I remember we tried to use global `tol` fixtures for closeness checks...?

File: tests/tape/interfaces/test_qnode_jax.py [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+    assert np.allclose(res, 1.8244501889992706, atol=tol, rtol=0)
+    assert np.allclose(grad[0], [-0.26610258, -0.47053553], atol=tol, rtol=0)
+    assert np.allclose(grad[1], 0.06486032, atol=tol, rtol=0)
```
Comment (mariaschuld) [inline tests/tape/interfaces/test_qnode_jax.py:205]: Would it be more unittest-like to just apply the transform and check the queue of the new qnode, rather than using a very elaborate post-processing function and checking some pre-computed numerical values?

**Review by mariaschuld — COMMENTED** [review]
Just left some comments for now. Awesome addition!

File: pennylane/tape/interfaces/jax.py (hunk 5 above)
Comment (Thenerdstation) [reply pennylane/tape/interfaces/jax.py:83]: Yes.

File: tests/tape/interfaces/test_qnode_jax.py (hunk 11 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py:33]: DeviceArray is the standard JAX tensor type but it isn't possible to use `isinstance(x, jax.DeviceArray)` sadly. This was the easiest workaround I could think of

File: tests/tape/interfaces/test_qnode_jax.py (hunk 14 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py:79]: You can't use backprop with a non-jax device.

File: tests/tape/interfaces/test_qnode_jax.py (hunk 15 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py:123]: I don't understand what you mean.

File: tests/tape/interfaces/test_qnode_jax.py (hunk 17 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py]: Yeah this comment should be deleted. new_circuit was treated as a global variable before and checked outside this method. However, we don't want to do that (the similar checks for the autograd interface don't apply here). I'll delete the comment.

File: tests/tape/interfaces/test_qnode_jax.py (hunk 19 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py:199]: I would disagree. The purpose of this check is to ensure that the `jax.grad` transform executes properly with the tape transform. The thing I specifically want to only check here is the numerics. If we only check the queue, it will miss the code path I'm aiming at.

File: tests/tape/interfaces/test_qnode_jax.py (hunk 13 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py]: Agreed, I should add more checks in this test.

File: tests/devices/test_default_qubit_jax.py (hunk 7 above)
Comment (Thenerdstation) [reply tests/devices/test_default_qubit_jax.py:96]: If I don't include it the tests fails. :(

File: tests/devices/test_default_qubit_jax.py (hunk 8 above)
Comment (Thenerdstation) [reply tests/devices/test_default_qubit_jax.py:94]: This was an early integration test that I think is pretty well covered by  the tests in `test_qnode_jax`. I'll remove it.

File: tests/devices/test_default_qubit_jax.py (hunk 9 above)
Comment (Thenerdstation) [reply tests/devices/test_default_qubit_jax.py:107]: Missed copy and paste error. I'll fix it.

File: tests/tape/interfaces/test_qnode_jax.py (hunk 16 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py:161]: Thanks, I stole if from josh's tests lol

**Comment by Thenerdstation** [comment]
@josh146 this should be ready to be merged

File: tests/tape/interfaces/test_qnode_jax.py (hunk 12 above)
Comment (Thenerdstation) [reply tests/tape/interfaces/test_qnode_jax.py:22]: Added!

File: tests/devices/test_default_qubit_jax.py (hunk 10 above)
Comment (Thenerdstation) [reply tests/devices/test_default_qubit_jax.py:458]: :)

File: pennylane/tape/interfaces/jax.py (hunk 5 above)
Comment (Thenerdstation) [reply pennylane/tape/interfaces/jax.py:83]: The reason I don't have this supported yet is because JAX requires to know the output shape of any `host_callback` call _before_ calling the function. Thankfully, we can determine the shape just from the tape, but it's complicated and I wanted to get a minimal version working before diving in.

File: tests/tape/interfaces/test_qnode_jax.py (hunk 14 above)
Comment (mariaschuld) [reply tests/tape/interfaces/test_qnode_jax.py:83]: True, at least not if we use the `mixed` device...

File: tests/tape/interfaces/test_qnode_jax.py (hunk 15 above)
Comment (mariaschuld) [reply tests/tape/interfaces/test_qnode_jax.py:127]: Oh I think I understand now on my side...you are not testing the expansion itself in this sense...all good.

**Comment by mariaschuld** [comment]
Overwriting the codecov because untested lines are falsely reported due to how jitting works.
